src/wasm/wasm-binary.cpp

```diff
@@ -330,18 +330,82 @@ void WasmBinaryWriter::writeExports() {
 
 void WasmBinaryWriter::writeDataSegments() {
   if (wasm->memory.segments.size() == 0) return;
-  uint32_t num = 0;
+  Index numConstant = 0,
+        numDynamic = 0;
   for (auto& segment : wasm->memory.segments) {
-    if (segment.data.size() > 0) num++;
+    if (segment.data.size() > 0) {
+      if (segment.offset->is<Const>()) {
+        numConstant++;
+      } else {
+        numDynamic++;
+      }
+    }
+  }
+  // check if we have too many dynamic data segments, which we can do nothing about
+  auto num = numConstant + numDynamic;
+  if (numDynamic + 1 >= MaxDataSegments) {
+    std::cerr << "too many non-constant-offset data segments, wasm VMs may not accept this binary" << std::endl;
+  }
+  // we'll merge constant segments if we must
+  if (numConstant + numDynamic >= MaxDataSegments) {
+    numConstant = MaxDataSegments - numDynamic - 1;
+    num = numConstant + numDynamic;
+    assert(num == MaxDataSegments - 1);
```

Comment:
Are there runtime cases where this might not be true?

I secretly like this as-is as an affirmation of the expected postcondition of the previous two lines, just in case the reader doesn't feel like doing the algebra. (Although my first reaction was "wait isn't this always true")

Reply:
Yes, mathematically speaking it should be true anyhow, but I think it's nice for clarity, and to prevent future refactorings from breaking stuff.